components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/mutualSSL_handler.bal

```diff
@@ -58,98 +58,155 @@ public type MutualSSLHandler object {
+                    } else { // if client certificate validation enabled for the certificate present in context
+                        //and header is also present then both should be validated.
+                        //This is the scenario where both client certificate is also should be verified and mtls is
+                        //also enabled between mgw and LB. So both client certificate present in the header should be
+                        //validated and the LB certificate present in the context.
+                        // When validating the certificate in the context we do not need to validate it with the alias
+                        //list present in the config as this would always be the LB certificate.
+                        mutualSSLStatus =  self.checkCertificatePresentInContext(req, apiName, apiVersion, false);
+                        if(mutualSSLStatus is boolean && mutualSSLStatus) {
+                            mutualSSLStatus = self.checkCertificatePresentInHeader(req, apiName, apiVersion);
                         }
                     }
+                } else {
+                //If certificate not in the header, and mutual ssl client validation config is enabled, then
+                // checking the certificate in request context is mandatory.(This case is when there is no LB fronted.)
+                //And also cert should be validated against with the API alias list. This is because the certificate
+                //available via the context would be the client certificate,not the LB one Hence the
+                //'isValidateCertificateWithAPI' value is set as true.
+                    mutualSSLStatus =  self.checkCertificatePresentInContext(req, apiName, apiVersion, true);
                 }
-                string? cert = req.mutualSslHandshake["base64EncodedCert"];
-                var cacheKey = cert.toString() + apiName + apiVersion;
-                var isExistCertCache = self.gatewayCache.retrieveFromMutualSslCertificateCache(cacheKey);
-                if (isExistCertCache is boolean)    {
-                    if (!isExistCertCache) {
-                        printDebug(KEY_AUTHN_FILTER,"Mutual SSL authentication failure. " +
-                        "API is not associated with the certificate");
-                        return false;
-                     }
+                if(mutualSSLStatus is boolean && mutualSSLStatus) {
+                    printDebug(KEY_AUTHN_FILTER, "MutualSSL handshake status: PASSED");
+                    doMTSLFilterRequest(req, invocationContext);
                 } else {
-                    handle|error certificateAlias = getAliasFromRequest(cert.toString());
-                    if (certificateAlias is error) {
-                        setErrorMessageToInvocationContext(API_AUTH_GENERAL_ERROR);
-                        return prepareAuthenticationError("Unclassified Authentication Failure");
+                    return mutualSSLStatus;
```

Comment:
I mutualSSLStatus is auth error or false the it would come here. Before returning false or auth error type from checkCertificatePresentInContext and checkCertificatePresentInHeader methods we set the response to invocation context